### ISSUE TITLE
RFE: non-const way to access menu items, similar to Fl_Group::child(int)

### ISSUE THREAD

**Issue by erco77** [fltk/fltk#314]
Currently to access a menu item to make changes (e.g. to the 'flags'), one has to do a bit of convoluted casting and cast-away-const:

    Fl_Choice choice(..);
    choice.add("Aaa");
    choice.add("Bbb");
    int index = choice.add("Ccc");
    ((Fl_Menu_Item*)&(choice.menu()[index]))->flags |= FL_MENU_INVISIBLE; // UGH

..which is based on `Fl_Menu_.cxx's` own code, BTW.

It would be nice if there was an easier way to access the items to make changes, e.g.

    choice.item(index)->flags |= FL_MENU_INVISIBLE; 

..or some such less painful method.

 I haven't researched if this can cause problems in some way, as I assume our `const menu()` method is `const` for a reason, but I'm guessing it goes back to xforms compatibility or some such. But there's certainly many real-world reasons for being able to modify menu items on the fly at runtime, and it should be easier to do.

**Comment by Albrecht-S** [comment]
I agree that such a method would be convenient and I do not object, but ...

> I haven't researched if this can cause problems in some way, as I assume our const menu() method is const for a reason, ...

The problem is that you can construct menus in static memory and write access to such menu items as proposed would be an access violation. OTOH, the way you create the menu in the example allocates it on the heap which is OK. Another way would be to copy the menu which is automatically done by some methods, e.g. adding a menu item, and then it's OK to access the copied menu item. I believe most of this is documented, but maybe not all consequences. Off the top of my head I can't tell which methods would copy the (entire) menu and how (or if) this is all documented.

That said, if we implemented such a method, this would all need to be documented clearly. If you cast away the constness it's your fault. If FLTK lets you access the menu items as non-const, it's FLTK's fault. Unless this is clearly documented.

Disclaimer: I did neither check the code nor the docs, it's just what I remember.

**Comment by erco77** [comment]
Good idea -- I'll be sure to mention the `const static` issue in the docs.

IIRC, just the act of add()ing an item to a static menu makes an in-memory copy which then makes it writable. There is also an Fl_Menu_::copy() function, not sure if that can be a workaround too.

Looking at Fl_Menu_.cxx, there is an 'owner' variable that seems to be in place for this purpose(?), but it's defined /below/ the copy() function, so not sure how that gets updated if a copy is made.. hrm.

I'll research it when I dive into this issue; I haven't yet.

**Comment by Albrecht-S** [comment]
> Looking at Fl_Menu_.cxx, there is an 'owner' variable that seems to be in place for this purpose(?), but it's defined /below/ the copy() function, so not sure how that gets updated if a copy is made.. hrm.

If (when) you do that: ISTR that there are issues WRT copying menus where it's not clear under which conditions the menu is owned by whom and can be destroyed when the widget is deleted. I'm not sure if there's an old STR (didn't find one in a quick search), it might have been a discussion in fltk.cordev (or general).

**Comment by Albrecht-S** [comment]
I found the discussion in fltk.coredev in thread "[Fl_Menu_::copy() just evil, what to do?](https://groups.google.com/g/fltkcoredev/c/eNsWc1KXv3Y/m/OyvnIXQ5_3AJ)"

@erco77 FYI, I just copied the link for reference, I didn't (and will not) investigate further.

**Comment by erco77** [comment]
Yikes, that sounds like it needs its own issue for sure to investigate.

I'm not going to get into that whole mess, that's a huge rabbit hole that needs its own army to investigate.

I may poke at it a little bit just to see, but in general I've never used the static array approach (though I know fluid does). Perhaps fluid should be changed or have an option to define menus as code (e.g. add()) rather than static arrays.

**Comment by erco77** [comment]
Opened a separate issue #332 regarding researching that Pierre Ossman thread.

**Comment by Albrecht-S** [comment]
> Looking at Fl_Menu_.cxx, there is an 'owner' variable ... defined /below/ the copy() function, so ...
I'll research it when I dive into this issue

@erco77 Is this `'owner'` variable you mentioned `'fl_menu_array_owner'`? I can't see another 'owner' variable. If yes, FWIW: this global variable is used to build a temporary array of menuitems while e.g. menuitems are `add()`ed or `insert()`ed.

**Comment by MatthiasWM** [comment]
Sigh, yes, Fl_Menu_Item arrays are an issue. C++11 does not have functions yet to check how those were allocated and if they are writable (introduced in C++20 and 23 IIRC).

So what I can offer is a new flag in Fl_Menu_Item::flags that indicates that the menu is a copy. It would be set only in the very last menu item where the label text is `nullptr` anyway. That would probably require a second flag to indicate if the user owns the array, or the Fl_Menu_ widget. 

Anyway, if that bit is set, the non-const functions work. Otherwise they fail with an assert. It would make access slow because we need to find the last item on every write access. Or we set those flags on *all* items that were copied.

Kludge on top of kludge...  ;-)

**Comment by Albrecht-S** [comment]
Long story short: the old code based on static menu item arrays that can be modified dynamically by code (add()) etc. is "broken by nature". It works as long as you know what you do.

IMHO we really need a new menu code base with a proper hierarchy that can be used in the future, i.e. by new applications. The old code should then probably be deprecated...

**Comment by MatthiasWM** [comment]
> IMHO we really need a new menu code base with a proper hierarchy that can be used in the future, i.e. by new applications. The old code should then probably be deprecated...

Good. I fully agree, and I am working on it already anyway.

**Comment by MatthiasWM** [comment]
Current state is to no longer add features to Fl_Menu_Item in the hopes that we get a better menu system working. If there are no additions to this issue, I will close it on Dec 13th.

**Comment by MatthiasWM** [comment]
Closing this, hoping to get to teh "new" menu implementation soon.
